### PR TITLE
Convert NativeImage instances to OS-specific representations

### PR DESCRIPTION
**Pull request by paulcbetts**
This PR adds a new `asNativeRepresentation` method to `NativeImage` which allows you to get an underlying image object that you can pass to native APIs. Right now only OS X is implemented but I've set it up so that we can add other operating systems later (others aren't so easy because Windows has more than one image representation format, etc etc)

### CODE REVIEW

File: atom/common/api/atom_api_native_image.cc [hunk 1]
```diff
@@ -221,6 +237,50 @@ std::string NativeImage::ToDataURL() {
...
 
+v8::Local<v8::Value> NativeImage::AsNativeRepresentation(
+    v8::Isolate* isolate,
+    mate::Arguments* args) {
+  NativeRepresentation desiredRep = NativeRepresentation::INVALID;
+  void* ptr = nullptr;
+  std::string type;
+
+  if (!args->GetNext(&type)) {
+    args->ThrowError();
```
Comment (paulcbetts) [inline atom/common/api/atom_api_native_image.cc]: This case crashes the renderer, but debugging it on OS X doesn't seem to help me much, I just get a bogus stack

File: atom/common/api/atom_api_native_image.cc [hunk 2]
```diff
@@ -184,6 +198,8 @@ mate::ObjectTemplateBuilder NativeImage::GetObjectTemplateBuilder(
     template_.Reset(isolate, mate::ObjectTemplateBuilder(isolate)
         .SetMethod("toPng", &NativeImage::ToPNG)
         .SetMethod("toJpeg", &NativeImage::ToJPEG)
+        .SetMethod("asNativeRepresentation",
```
Comment (zcbenz) [inline atom/common/api/atom_api_native_image.cc]: I prefer `getNativeHandle` as API name, which matches the style of `BrowserWindow.getNativeWindowHandle`.

File: atom/common/api/atom_api_native_image.cc [hunk 3]
```diff
@@ -221,6 +237,50 @@ std::string NativeImage::ToDataURL() {
...
+v8::Local<v8::Value> NativeImage::AsNativeRepresentation(
+    v8::Isolate* isolate,
+    mate::Arguments* args) {
+  NativeRepresentation desiredRep = NativeRepresentation::INVALID;
+  void* ptr = nullptr;
+  std::string type;
+
+  if (!args->GetNext(&type)) {
+    args->ThrowError();
+    goto out;
```
Comment (zcbenz) [inline atom/common/api/atom_api_native_image.cc]: After throwing error, we should `return v8::Undefined(isolate)`, the `goto` is probably responsible for the crash.

File: docs/api/native-image.md [hunk 4]
```diff
@@ -133,6 +133,13 @@ Returns a [Buffer][buffer] that contains the image's `JPEG` encoded data.
 
 Returns the data URL of the image.
 
+### `image.asNativeRepresentation(type)`
+
+Returns a pointer to an underlying native type (encoded as a [Buffer][buffer]) which can be used with native APIs. Note that in many cases, this pointer is a weak pointer to the underlying native image not a copy, so you _must_ ensure that the associated `nativeImage` instance is kept around.
+
+* `type` String (**required**) - one of:
```
Comment (zcbenz) [inline docs/api/native-image.md]: The `type` parameter seems redundant to me, since on each platform there is only one representation of native image.

Comment (paulcbetts) [inline docs/api/native-image.md]: So on Windows you have at least `HICON` and `HBITMAP`, and I imagine there might be more depending on what people want

File: atom/common/api/atom_api_native_image.cc (hunk 2 above)
Comment (paulcbetts) [inline atom/common/api/atom_api_native_image.cc]: :+1:

File: docs/api/native-image.md (hunk 4 above)
Comment (zcbenz) [inline docs/api/native-image.md]: On Linux and Window the native images are always represented as bitmaps in Chromium under the hood, if users want to get the image as `HBITMAP` or other types of native resources, a conversation has to be made, which makes it more like a `toXXX` API.

Comment (paulcbetts) [inline docs/api/native-image.md]: Legit

File: spec/api-native-image-spec.js [hunk 5]
```diff
@@ -34,5 +34,18 @@ describe('nativeImage module', () => {
...
+      if (process.platform !== 'darwin') return;
+
+      const imagePath = `${path.join(__dirname, 'fixtures', 'api')}${path.sep}..${path.sep}${path.join('assets', 'logo.png')}`;
+      const image = nativeImage.createFromPath(imagePath);
+      const nsimage = image.getNativeHandle();
+
+      assert.equal(nsimage.length, 8);
+
+      // If all bytes are null, that's Bad
+      assert.equal(nsimage.reduce((acc,x) => acc || (x != 0)), true);
```
Comment (zcbenz) [inline spec/api-native-image-spec.js]: This test is failing: `AssertionError: 152 == true`.

File: atom/common/api/atom_api_native_image.cc [hunk 6]
```diff
@@ -221,6 +223,22 @@ std::string NativeImage::ToDataURL() {
   return data_url;
 }
 
+v8::Local<v8::Value> NativeImage::GetNativeHandle(
+    v8::Isolate* isolate,
+    mate::Arguments* args) {
+#if defined(OS_MACOSX)
+  void* ptr = reinterpret_cast<void*>(image_.AsNSImage());
+#else
+  args.ThrowError();
```
Comment (zcbenz) [inline atom/common/api/atom_api_native_image.cc]: Compilation is failing on platforms other than OS X.

**Comment by zcbenz** [comment]
:+1:

**Pull request merged**